packages/openapi3/src/decorators.ts

```diff
@@ -32,3 +50,64 @@ export const $oneOf: OneOfDecorator = (
 export function getOneOf(program: Program, entity: Type): boolean {
   return program.stateMap(oneOfKey).get(entity);
 }
+
+const [getTagsMetadata, setTagsMetadata] = unsafe_useStateMap<
+  Type,
+  { [name: string]: OpenAPI3Tag }
+>(OpenAPI3Keys.tagsMetadata);
+
+export const $tagMetadata: TagMetadataDecorator = (
+  context: DecoratorContext,
+  entity: Namespace,
+  name: string,
+  tagMetadata?: TypeSpecValue,
+) => {
+  const tags = getTagsMetadata(context.program, entity);
+  if (tags && tags[name]) {
+    reportDiagnostic(context.program, {
+      code: "duplicate-tag",
+      format: { tagName: name },
+      target: context.getArgumentTarget(0)!,
+    });
+  }
+  let metadata: OpenAPI3Tag = { name };
+  if (tagMetadata) {
+    const [data, diagnostics] = typespecTypeToJson<OpenAPI3Tag & Record<ExtensionKey, unknown>>(
+      tagMetadata,
+      context.getArgumentTarget(0)!,
+    );
+    context.program.reportDiagnostics(diagnostics);
+    if (data === undefined) {
+      return;
+    }
+    validateAdditionalInfoModel(context, tagMetadata, data);
+    metadata = { ...data, name };
+  }
+
+  const newTags = { ...tags, [name]: metadata };
+  setTagsMetadata(context.program, entity, newTags);
```

Comment:
```suggestion
  const newTags = { ...tags, [name]: metadata };
  setTagsMetadata(context.program, entity, newTags);
```
tags[name] = metadata;